ui/org.eclipse.pde.genericeditor.extension/src/org/eclipse/pde/internal/genericeditor/target/extension/model/RepositoryCache.java

```diff
@@ -31,41 +42,73 @@
-
-	Map<String, List<UnitNode>> cache = new HashMap<>();
-
 	private RepositoryCache() {
 		//avoid instantiation
 	}
 
-	/**
-	 * @return default instance of this cache.
-	 */
-
-	public static RepositoryCache getDefault() {
-		if (instance == null) {
-			instance = new RepositoryCache();
-		}
-		return instance;
-	}
+	private static final Map<String, CompletableFuture<Map<String, List<IVersionedId>>>> CACHE = new ConcurrentHashMap<>();
 
 	/**
 	 * Fetches information and caches it.
-	 *
-	 * @param repo
-	 *            repository URL
-	 * @param flush
-	 *            whether a flush is needed
-	 * @return list of IUs available in the 'repo' repository. Never
-	 *         <code>null</code>.
+	 * <p>
+	 * All available IUs are returned as a map mapping the IDs of all IUs
+	 * available in the {@code repositories} to the list of all available
+	 * {@link IVersionedId versioned IDs} for that ID. All keys are sorted in
+	 * alphabetical order and all versions are sorted in descending order.
+	 * </p>
+	 * 
+	 * @return all available units in the specified {@code repository} in a map
+	 *         mapping all IDs to all available versions.
 	 */
-	public List<UnitNode> fetchP2UnitsFromRepo(String repo, boolean flush) {
-		if ((flush) || (cache.get(repo) == null)) {
-			List<UnitNode> units = P2Fetcher.fetchAvailableUnits(repo);
-			cache.put(repo, units);
+	public static Map<String, List<IVersionedId>> fetchP2UnitsFromRepos(List<String> repositories) {
+		if (repositories.size() == 1) {
+			try {
+				return fetchP2DataOfRepo(repositories.get(0)).get();
+			} catch (InterruptedException | ExecutionException e) {
+				return Map.of();
+			}
 		}
-		return cache.get(repo);
+		List<Future<Map<String, List<IVersionedId>>>> repos = repositories.stream()
+				.map(RepositoryCache::fetchP2DataOfRepo).toList();
+		// Fetch all repos at once to await pending metadata in parallel
+		return toSortedMap(repos.stream().<Map<String, List<IVersionedId>>>map(f -> {
+			try {
+				return f.get();
+			} catch (InterruptedException | ExecutionException e) {
+				return Map.of();
+			}
+		}).map(Map::values).flatMap(Collection::stream).flatMap(List::stream));
+	}
+
+	public static void prefetchP2MetadataOfRepository(String repository) {
+		fetchP2DataOfRepo(repository);
+	}
+
+	private static Future<Map<String, List<IVersionedId>>> fetchP2DataOfRepo(String repository) {
+		return CACHE.compute(repository, (r, f) -> {
+			if (f != null && f.isDone() && !f.isCompletedExceptionally() && !f.isCancelled()) {
+				return f;
+			}
+			CompletableFuture<Map<String, List<IVersionedId>>> future = new CompletableFuture<>();
+			// Fetching P2 repository information is a costly operation
+			// time-wise. Thus it is done in a job.
+			Job job = Job.create(NLS.bind(Messages.UpdateJob_P2DataFetch, r), m -> {
+				Map<String, List<IVersionedId>> map = toSortedMap(P2Fetcher.fetchAvailableUnits(r));
+				future.complete(map);
+			});
+			job.setUser(true);
```

Comment:
I think Jobs should have a group here as there could be many of them.

Reply:
I tried that but in the UI I couldn't see a difference and in code it's not needed, since the joint completion is handled via the Future. Therefore I cannot see a real advantage here.

What also should be considered that at maximum all declared repos of a location are queried in parallel.
So I assume the number of parallel jobs not to be too great.